kv/kvserver: TestLeaseQueueProactiveEnqueueOnPreferences failed

kv/kvserver.TestLeaseQueueProactiveEnqueueOnPreferences [failed](https://tanzanite.cluster.engflow.com/invocations/default/11e796f0-9570-45c5-9c5b-813c999756ca?testReportRun=3&testReportShard=22&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [b5081e997adaf5950c9b85dfb6655f226bf16d29](https://github.com/cockroachdb/cockroach/commits/b5081e997adaf5950c9b85dfb6655f226bf16d29):


```
=== RUN   TestLeaseQueueProactiveEnqueueOnPreferences
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestLeaseQueueProactiveEnqueueOnPreferences3796292043
    test_log_scope.go:81: use -show-logs to present logs inline
    lease_queue_test.go:410: -- test log scope end --
test logs left over in: outputs.zip/logTestLeaseQueueProactiveEnqueueOnPreferences3796292043
--- FAIL: TestLeaseQueueProactiveEnqueueOnPreferences (56.73s)
=== RUN   TestLeaseQueueProactiveEnqueueOnPreferences/violating
    lease_queue_test.go:383: condition failed to evaluate within 45s: from lease_queue_test.go:389: lease not on preferred node 1, on n3,s3
    --- FAIL: TestLeaseQueueProactiveEnqueueOnPreferences/violating (45.20s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=3</code>
 - <code>shard=22</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #124691 kv/kvserver: TestLeaseQueueProactiveEnqueueOnPreferences failed [C-test-failure O-robot T-kv branch-release-24.1 release-blocker]
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeaseQueueProactiveEnqueueOnPreferences.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-39013

kv/kvserver.TestLeaseQueueProactiveEnqueueOnPreferences [failed](https://tanzanite.cluster.engflow.com/invocations/default/b801cc73-460d-4704-9f2a-e615be2e7273?testReportRun=1&testReportShard=22&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [bd4c5fd7c65c0a9afa8ba05675e49747a4f68881](https://github.com/cockroachdb/cockroach/commits/bd4c5fd7c65c0a9afa8ba05675e49747a4f68881):


```
=== RUN   TestLeaseQueueProactiveEnqueueOnPreferences
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestLeaseQueueProactiveEnqueueOnPreferences2520501900
    test_log_scope.go:81: use -show-logs to present logs inline
    lease_queue_test.go:410: -- test log scope end --
test logs left over in: outputs.zip/logTestLeaseQueueProactiveEnqueueOnPreferences2520501900
--- FAIL: TestLeaseQueueProactiveEnqueueOnPreferences (56.39s)
=== RUN   TestLeaseQueueProactiveEnqueueOnPreferences/less-preferred
    lease_queue_test.go:398: condition failed to evaluate within 45s: from lease_queue_test.go:404: lease not on preferred node 1, on n2,s2
    --- FAIL: TestLeaseQueueProactiveEnqueueOnPreferences/less-preferred (45.20s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=1</code>
 - <code>shard=22</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #124691 kv/kvserver: TestLeaseQueueProactiveEnqueueOnPreferences failed [C-test-failure O-robot P-1 T-kv X-duplicate branch-release-24.1 release-blocker]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeaseQueueProactiveEnqueueOnPreferences.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

--- a/pkg/kv/kvserver/liveness/liveness.go
+++ b/pkg/kv/kvserver/liveness/liveness.go
@@ -701,7 +701,9 @@ var errNodeAlreadyLive = errors.New("node already live")
 //
 // If this method returns nil, the node's liveness has been extended,
 // relative to the previous value. It may or may not still be alive
-// when this method returns.
+// when this method returns. It may also not have been extended as far
+// as the livenessThreshold, because the caller may have raced with
+// another heartbeater.
 //
 // On failure, this method returns ErrEpochIncremented, although this
 // may not necessarily mean that the epoch was actually incremented.
@@ -839,17 +841,6 @@ func (nl *NodeLiveness) heartbeatInternal(
 		// expired while in flight, so maybe we don't have to care about
 		// that and only need to distinguish between same and different
 		// epochs in our return value.
-		//
-		// TODO(nvanbenschoten): Unlike the early return above, this doesn't
-		// guarantee that the resulting expiration is past minExpiration,
-		// only that it's different than our oldLiveness. Is that ok? It
-		// hasn't caused issues so far, but we might want to detect this
-		// case and retry, at least in the case of the liveness heartbeat
-		// loop. The downside of this is that a heartbeat that's intending
-		// to bump the expiration of a record out 9s into the future may
-		// return a success even if the expiration is only 5 seconds in the
-		// future. The next heartbeat will then start with only 0.5 seconds
-		// before expiration.
 		if actual.IsLive(nl.clock.Now()) && !incrementEpoch {
 			return errNodeAlreadyLive
 		}

--- a/pkg/kv/kvserver/replica_range_lease.go
+++ b/pkg/kv/kvserver/replica_range_lease.go
@@ -538,20 +538,31 @@ func (p *pendingLeaseRequest) requestLease(
 	// interval are the same.
 	expToEpochPromo := extension && status.Lease.Type() == roachpb.LeaseExpiration && reqLease.Type() == roachpb.LeaseEpoch
 	if expToEpochPromo && reqLeaseLiveness.Expiration.ToTimestamp().Less(status.Lease.GetExpiration()) {
-		err := p.repl.store.cfg.NodeLiveness.Heartbeat(ctx, reqLeaseLiveness)
-		if err != nil {
-			if logFailedHeartbeatOwnLiveness.ShouldLog() {
-				log.Errorf(ctx, "failed to heartbeat own liveness record: %s", err)
+		curLiveness := reqLeaseLiveness
+		for r := retry.StartWithCtx(ctx, base.DefaultRetryOptions()); r.Next(); {
+			err := p.repl.store.cfg.NodeLiveness.Heartbeat(ctx, curLiveness)
+			if err != nil {
+				if logFailedHeartbeatOwnLiveness.ShouldLog() {
+					log.Errorf(ctx, "failed to heartbeat own liveness record: %s", err)
+				}
+				return kvpb.NewNotLeaseHolderError(roachpb.Lease{}, p.repl.store.StoreID(), p.repl.Desc(),
+					fmt.Sprintf("failed to manipulate liveness record: %s", err))
 			}
-			return kvpb.NewNotLeaseHolderError(roachpb.Lease{}, p.repl.store.StoreID(), p.repl.Desc(),
-				fmt.Sprintf("failed to manipulate liveness record: %s", err))
-		}
-		// Assert that the liveness record expiration is now greater than the
-		// expiration of the lease we're promoting.
-		l, ok := p.repl.store.cfg.NodeLiveness.GetLiveness(reqLeaseLiveness.NodeID)
-		if !ok || l.Expiration.ToTimestamp().Less(status.Lease.GetExpiration()) {
-			return errors.AssertionFailedf("expiration of liveness record %s is not greater than "+
-				"expiration of the previous lease %s after liveness heartbeat", l, status.Lease)
+			// Check whether the liveness record expiration is now greater than the
+			// expiration of the lease we're promoting. If not, we may have raced with
+			// another liveness heartbeat which did not extend the liveness expiration
+			// far enough and we should try again.
+			l, ok := p.repl.store.cfg.NodeLiveness.GetLiveness(reqLeaseLiveness.NodeID)
+			if !ok {
+				return errors.NewAssertionErrorWithWrappedErrf(liveness.ErrRecordCacheMiss, "after heartbeat")
+			}
+			if l.Expiration.ToTimestamp().Less(status.Lease.GetExpiration()) {
+				log.Infof(ctx, "expiration of liveness record %s is not greater than "+
+					"expiration of the previous lease %s after liveness heartbeat, retrying...", l, status.Lease)
+				curLiveness = l.Liveness
+				continue
+			}
+			break
 		}
 	}
 

